src/wal.rs

```diff
@@ -100,69 +94,246 @@ pub struct Wal {
+        ))
     }
 
-    pub fn truncate(&mut self, _end: u64) -> Result<()> {
-        unimplemented!()
+    /// Truncate WAL
+    pub fn truncate(&mut self, end: u64) -> Result<()> {
+        // TODO: check read only
+        let metadata = self.file.metadata()?;
+        if metadata.len() == end {
+            return Ok(());
+        }
+        self.size = end as u32;
+        self.file.set_len(end)?;
+        Ok(())
     }
 
-    fn done_writing(&mut self, _offset: u32) -> Result<()> {
-        unimplemented!()
+    /// Finish WAL writing
+    pub(crate) fn done_writing(&mut self, offset: u32) -> Result<()> {
+        if self.opts.sync_writes {
+            self.file.sync_all()?;
+        }
+        self.truncate(offset as u64)?;
+        Ok(())
     }
 
+    /// Get WAL iterator
     pub fn iter(&mut self) -> Result<WalIterator> {
-        unimplemented!()
+        Ok(WalIterator::new(Cursor::new(
+            &self.mmap_file[0..self.size as usize],
+        )))
+    }
+
+    pub fn should_flush(&self) -> bool {
+        self.write_at as u64 > self.opts.value_log_file_size
+    }
+
+    /// Get real size of WAL. After truncating, WAL mmap file will have different
+    /// size against real file. `size` stores the actual length.
+    pub(crate) fn size(&self) -> u32 {
+        self.size
+    }
+
+    /// When using WAL as value log, we will need to extend or shrink actual size
+    /// of WAL file from outside functions.
+    pub(crate) fn set_size(&mut self, size: u32) {
+        self.size = size;
+    }
+
+    /// When using WAL as value log, we will need to extend or shrink actual size
+    /// of WAL file from outside functions.
+    pub(crate) fn set_len(&mut self, len: u64) -> Result<()> {
+        self.file.set_len(len)?;
+        Ok(())
+    }
+
+    pub(crate) fn data(&mut self) -> &mut MmapMut {
+        &mut self.mmap_file
     }
 }
 
 pub struct WalIterator<'a> {
-    reader: BufReader<&'a mut File>,
+    /// `reader` stores the file to read
+    reader: Cursor<&'a [u8]>,
+    /// `entry_reader` operates on `reader` and buffers entry information
     entry_reader: EntryReader,
 }
 
 impl<'a> WalIterator<'a> {
-    pub fn new(_reader: BufReader<&'a mut File>) -> Self {
-        unimplemented!()
+    pub fn new(reader: Cursor<&'a [u8]>) -> Self {
+        Self {
+            reader,
+            entry_reader: EntryReader::new(),
+        }
     }
 
+    /// Get next entry from WAL
     pub fn next(&mut self) -> Option<Result<EntryRef<'_>>> {
-        unimplemented!()
+        use std::io::ErrorKind;
+
+        let entry = self.entry_reader.entry(&mut self.reader);
+
+        match entry {
+            Ok(entry) => {
+                if entry.is_zero() {
+                    return None;
+                }
+                // TODO: process transaction-related metadata
+                Some(Ok(entry))
+            }
+            Err(Error::Io(err)) => {
+                if err.kind() == ErrorKind::UnexpectedEof {
```

Comment:
What if it reads a truncated entry?

Reply:
I have added tests for reading truncated entry.

There're 3 kinds of error when we read a truncated entry.

* If header is less than 2 bytes, we could not read meta. There will be `VarDecode` error.
* If varint key length and value length is truncated, prost will return `Decode` error.
* If key / value is truncated, `Cursor` will return `UnexpectedEof`.

Reply:
Then how to tell if it's truncated or just reach the end of the file?

Another question is why return `Option<Result>` instead of `Result<Option>`?

Reply:
`None` means a recoverable error (e.g. truncate, decode error), and `Some(Err)` refers to an unexpected error. Maybe it's better to exchange them.

And for the truncate issue, I'm a little bit confused about in what state a file is considered `truncate`. I'll look into this soon.

Reply:
From https://github.com/dgraph-io/badger/blob/master/memtable.go#L475 , we do not need to tell whether it's truncated or reached end of file. In both situation, we simply stop reading from the log file.

I've modified current implementation. Now it will:
* return `Ok(None)` if we reached first corrupted entry, and we could stop iteration
* return `Ok(Some(entry))` if we read a new entry
* return `Err` if the error is not recoverable